lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs

```diff
@@ -506,24 +524,35 @@ monitorStakePools tr gp nl db@DBLayer{..} = do
         -> (BlockHeader, ProtocolParameters)
         -> IO (FollowAction ())
     forward latestGarbageCollectionEpochRef blocks (_nodeTip, _pparams) = do
-        atomically $ forM_ blocks $ \case
-            BlockByron _ -> pure ()
-            BlockShelley blk -> do
-                let (slot, certificates) = poolCertsFromShelleyBlock blk
-                let header = toShelleyBlockHeader getGenesisBlockHash blk
-                runExceptT (putPoolProduction header (getProducer blk))
-                    >>= \case
-                        Left e ->
-                            liftIO $ traceWith tr $ MsgErrProduction e
-                        Right () ->
-                            pure ()
-                unless (null certificates) $ do
-                    -- Before adding new pool certificates to the database, we
-                    -- first attempt to garbage collect pools that have already
-                    -- retired.
-                    garbageCollectPools slot latestGarbageCollectionEpochRef
-                    putPoolCertificates slot certificates
+        atomically $ do
+            forM_ blocks $ \case
+                BlockByron _ -> pure ()
+                BlockShelley blk -> do
+                    let (slot, certificates) = poolCertsFromShelleyBlock blk
+                    let header = toShelleyBlockHeader getGenesisBlockHash blk
+                    handleErr (putPoolProduction header (getProducer blk))
+                    unless (null certificates) $ do
+                        -- Before adding new pool certificates to the database, we
+                        -- first attempt to garbage collect pools that have already
+                        -- retired.
+                        garbageCollectPools slot latestGarbageCollectionEpochRef
+                        putPoolCertificates slot certificates
+
+            -- We store the last byron header in the chunk only to have a sync point,
+            -- when syncing was interrupted.
+            case NE.last blocks of
+                BlockByron blk -> do
+                  dropByronHeaders
+                  handleErr (putByronHeader (toByronBlockHeader gp blk))
+                BlockShelley _ -> pure ()
```

Comment:
Though about it. But I think it might tempt someone to use two forM blocks by accident, which would be a performance penalty. This way we can see that we want to do as little processing as possible.

Reply:
In fact, we're already traversing twice due to `NE.last`... I'll try to fix that

Reply:
```hs
        -- | Like 'forM_', except runs the second action for the last element as
        -- well (in addition to the first action).
        forML_ :: (Monad m)
            => [a]
            -> (a -> m b) -- ^ action to run for all elements
            -> (a -> m c) -- ^ action to run for the last element
            -> m ()
        forML_ []  _ _ = pure ()
        forML_ [x] a1 a2 = a1 x >> a2 x >> forML_ [] a1 a2
        forML_ (x:xs) a1 a2 = a1 x >> forML_ xs a1 a2
```

Reply:
Surprisingly, this made it slower... maybe the other form has better list fusion/inlining and memoization.

Reply:
What if you write the _first_ byron block header instead of the _last_, does that help?

The checkpoint will be a bit old, but that shouldn't really matter I think.